metaflow/tracing_otel.py

```diff
@@ -0,0 +1,170 @@
+import os
+
+import sys
+from functools import wraps
+import contextlib
+from typing import Dict, List, Optional
+from opentelemetry import trace as trace_api, context
+from opentelemetry.trace.span import format_trace_id
+from opentelemetry.propagate import extract, inject
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
+    OTLPSpanExporter,
+)
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+
+tracer_provider = None
+
+
+def init_tracing():
+    from metaflow.metaflow_config import (
+        SERVICE_AUTH_KEY,
+        SERVICE_HEADERS,
+        OTEL_ENDPOINT,
+    )
+
+    global tracer_provider
+    if tracer_provider is not None:
+        print("Tracing already initialized", file=sys.stderr)
+        return
+    # print("initializing tracing", os.environ.get("traceparent"))
+
+    from metaflow.tracing_propagator import EnvPropagator
+
+    set_global_textmap(EnvPropagator(None))
+
+    if SERVICE_AUTH_KEY:
+        span_exporter = OTLPSpanExporter(
```

Comment:
Open discussion: Is it possible to make the propogator and exporter configurable?
For example, if the user want to use ZIPKIN, they're supposed to use `opentelemetry.propagators.b3.B3MultiFormat` propagator, and `opentelemetry.exporter.zipkin.proto.http.ZipkinExporter`, which may add multiple if/else in the script, something like:
```
if otel:
    use propagator A, exporter A
if zipkin:
    use propagator B, exporter B
if xxx:
    use propagotor C, exporter C
```
Just wondering if we can offer a cleaner way to configure propagator and exporter

Reply:
Generally OTEL SDK supports [doing this through environment variables](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/). Usually when instrumenting services, like web backends, I'd just rely on those. 

However in Metaflow case, its hard to expect end users to set all those correctly and making sure those are consistently set across all subproces/task boundaries is pretty error prone. So we opted for not using this mechanism to simplify the implementation. 

Maybe initially we could support a couple combinations that are most practically useful (lets say OTEL HTTP & Zipkin) via a metaflow config var, something like METAFLOW_OTEL_MODE=zipkin and come back to this later if more flexibility is necessary.